core/src/main/java/com/oath/oak/InternalOakMap.java

```diff
@@ -574,100 +615,141 @@ V remove(K key, V oldValue, Function<ByteBuffer, V> transformer) {
-            if (state == Chunk.State.INFANT) {
-                // the infant is already connected so rebalancer won't add this put
-                rebalance(c.creator());
-                continue;
-            }
-            if (state == Chunk.State.FROZEN || state == Chunk.State.RELEASED) {
-                if (!rebalance(c)) {
-                    continue;
-                }
-                return v;
-            }
-
-
             assert lookUp.entryIndex > 0;
             assert lookUp.valueReference != INVALID_VALUE_REFERENCE;
-            Chunk.OpData opData = new Chunk.OpData(Operation.REMOVE, lookUp.entryIndex, INVALID_VALUE_REFERENCE,
-                    lookUp.valueReference, null);
 
             // publish
-            if (!c.publish()) {
-                if (!rebalance(c)) {
-                    continue;
-                }
-                return v;
+            if (c.finalizeDeletion(lookUp)) {
+                rebalance(c);
             }
-
-            finishAfterPublishing(opData, c);
         }
     }
 
     OakRBuffer get(K key) {
         if (key == null) {
             throw new NullPointerException();
         }
-
-        Chunk<K, V> c = findChunk(key); // find chunk matching key
-        Chunk.LookUp lookUp = c.lookUp(key);
-        if (lookUp == null || lookUp.valueSlice == null) {
-            return null;
+        while (true) {
+            Chunk<K, V> c = findChunk(key); // find chunk matching key
+            Chunk.LookUp lookUp = c.lookUp(key);
+            if (lookUp == null || lookUp.valueSlice == null) {
+                return null;
+            }
+            if (updateVersionAfterLinking(c, lookUp)) {
+                continue;
+            }
+            long keyReference = c.getKeyReference(lookUp.entryIndex);
+            return new OakRValueBufferImpl(lookUp.valueReference, lookUp.version, keyReference, operator,
+                    memoryManager, this);
         }
-        return new OakRValueBufferImpl(lookUp.valueSlice.getByteBuffer().asReadOnlyBuffer());
     }
 
-    <T> T getValueTransformation(K key, Function<ByteBuffer, T> transformer) {
-        if (key == null || transformer == null) {
+    boolean computeIfPresent(K key, Consumer<OakWBuffer> computer) {
+        if (key == null || computer == null) {
             throw new NullPointerException();
         }
 
-        Chunk<K, V> c = findChunk(key); // find chunk matching key
-        Chunk.LookUp lookUp = c.lookUp(key);
-        if (lookUp == null || lookUp.valueSlice == null) {
-            return null;
+        while (true) {
+            Chunk<K, V> c = findChunk(key); // find chunk matching key
+            Chunk.LookUp lookUp = c.lookUp(key);
+
+            if (lookUp != null && lookUp.valueSlice != null) {
+                if (updateVersionAfterLinking(c, lookUp)) {
+                    continue;
+                }
+                NovaValueUtils.NovaResult res = operator.compute(lookUp.valueSlice, computer, lookUp.version);
```

Comment:
Operator appears to be the "handle" to make the operations on the existing values like compute/put/exchange/etc. However this is very opaque. Especially because it is one operator providing "all the services". It would be much clearer if we have "something" related to a specific value on which compute/put/exchange/etc. operations can be applied. It can be LookUp, Slice, ...